src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1147,13 +1147,23 @@ static BoundCollectionExpressionSpreadElement bindSpreadElement(
                         ErrorCode.ERR_CollectionArgumentsNotSupportedForType,
                         _node.WithElement.Syntax.GetFirstToken().GetLocation(),
                         _targetType);
+                    return null;
                 }
 
                 return CreateCollectionExpression(collectionTypeKind, elements);
             }
 
             private readonly BoundCollectionExpression? TryConvertCollectionExpressionArrayInterfaceType(ImmutableArray<BoundNode> elements)
             {
+                if (_node.WithElement?.Arguments.Length > 0 &&
+                    _targetType.IsReadOnlyArrayInterface(out _))
+                {
+                    // For the read-only array interfaces (IEnumerable<E>, IReadOnlyCollection<E>, IReadOnlyList<E>), only
+                    // the parameterless `with()` is allowed.
+                    _diagnostics.Add(ErrorCode.ERR_CollectionArgumentsMustBeEmpty, _node.WithElement.Syntax.GetFirstToken().GetLocation());
+                    return null;
```

Comment:
>return null; [](http://example.com/codeflow?start=20&length=12)

We should have a test verifying what happens with `WithElement` in IOperation tree when we hit this code path.